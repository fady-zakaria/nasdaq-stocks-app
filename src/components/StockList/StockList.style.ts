import { StyleSheet } from "react-native";
import { Colors } from "../../utils/Colors";

export const styles = StyleSheet.create({
    ContentContainer: {
        justifyContent: 'space-between',
        paddingVertical: 5,
    },
    emptyContainer: {
        padding: 10,
        alignItems: 'center',
    },
    emptyText: {
        color: Colors.primaryText,
        fontSize: 15,
    }
})
