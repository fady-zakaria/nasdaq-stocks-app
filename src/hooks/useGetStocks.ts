import { useInfiniteQuery } from '@tanstack/react-query';
import getStocks from '../services/getStocks.service';
import { useMemo } from 'react'

export interface QueryParam {
  active?: Boolean,
  limit?: number;
  search?: string;
}

export const useGetStocks = (queryParam: QueryParam) => {
  const queryKey = useMemo(
    () => [
      'stocks',
      queryParam.limit,
      queryParam.search,
    ],
    [queryParam.limit, queryParam.search],
  );
  const query = useInfiniteQuery({
    queryKey: [queryKey],
    queryFn: async ({ pageParam = '' }) => await getStocks(pageParam, queryParam),
    getNextPageParam: (lastPage) => lastPage.data.next_url,
    staleTime: 30000,
  });
  return {
    ...query,
  };
};
