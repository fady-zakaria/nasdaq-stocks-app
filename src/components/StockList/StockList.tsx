import { View, FlatList, StyleSheet } from 'react-native';
import React, { FC, useEffect, useState } from 'react';
import TickerBox from '../TickerBox/TickerBox';
import { useGetStocks } from '../../hooks/useGetStocks';
import { Ticker } from '../../types/stocks';
import Loader from '../Loader/Loader';

interface Iprops {
  searchQuery: string;
}

const StockList: FC<Iprops> = ({ searchQuery }) => {

  const {
    data,
    refetch,
    isRefetching,
    fetchNextPage,
    isFetchingNextPage,
    isLoading,
    isFetching,
  } = useGetStocks({
    active: true,
    limit: 10,
    search: searchQuery && searchQuery.length > 2 ? searchQuery : '',
  });

  const [queryRes, setQueryRes] = useState<Ticker[]>([]);

  useEffect(() => {
    let newArray = [];
    // console.log("data",data)
    data?.pages.map(item => {
      if (item) {
        newArray = [...newArray, ...item?.data?.results];
      }
    });
    setQueryRes(newArray);
  }, [data]);

  const renderItem = ({ item }) => {
    return (
      <TickerBox name={item.name} ticker={item.ticker} />
    )
  }

  return (
    <View>
      {isRefetching || isLoading ? <Loader /> : (
        <FlatList
          numColumns={2}
          data={queryRes}
          renderItem={renderItem}
          showsVerticalScrollIndicator={false}
          keyExtractor={(item, index) => (item.ticker + index).toString()}
          columnWrapperStyle={styles.ContentContainer}
          onEndReached={() => !isFetching && (setTimeout(() => {
            fetchNextPage();
          }, 1000))}
          ListFooterComponent={
            isFetchingNextPage ? (
              <Loader />
            ) : null
          }
          testID="stocksList"
        />
      )}
    </View>
  );
};

export default StockList;

const styles = StyleSheet.create({
  ContentContainer: {
    justifyContent: 'space-between',
    paddingVertical: 5,
  },
});
