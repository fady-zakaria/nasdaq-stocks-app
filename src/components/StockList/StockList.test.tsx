import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react-native';
import StockList from './StockList';
import { useGetStocks } from '../../hooks/useGetStocks';

jest.mock('../../hooks/useGetStocks');

const mockUseGetStocks = useGetStocks as jest.MockedFunction<typeof useGetStocks>;

describe('StockList', () => {
    beforeEach(() => {
        mockUseGetStocks.mockReturnValue({
            data: {
                pages: [
                    { data: { results: [{ name: 'Apple', ticker: 'AAPL' }, { name: 'Google', ticker: 'GOOGL' }] } },
                ],
            },
            refetch: jest.fn(),
            isRefetching: false,
            fetchNextPage: jest.fn(),
            isFetchingNextPage: false,
            isLoading: false,
            isFetching: false,
        });
    });

    it('renders correctly with given props', () => {
        const tree = renderer.create(
            <StockList searchQuery="test" />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('renders the loader when loading', () => {
        mockUseGetStocks.mockReturnValueOnce({
            data: null,
            refetch: jest.fn(),
            isRefetching: true,
            fetchNextPage: jest.fn(),
            isFetchingNextPage: true,
            isLoading: true,
            isFetching: false,
        });

        const tree = renderer.create(
            <StockList />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('renders the list of stocks correctly', () => {
        const { getByText } = render(<StockList searchQuery="test" />);

        expect(getByText('Apple')).toBeTruthy();
        expect(getByText('Google')).toBeTruthy();
    });
});
