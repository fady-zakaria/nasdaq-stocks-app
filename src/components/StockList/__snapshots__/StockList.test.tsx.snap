// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StockList renders correctly with given props 1`] = `
<View>
  <RCTScrollView
    ListFooterComponent={null}
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    testID="stocksList"
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</View>
`;

exports[`StockList renders the loader when loading 1`] = `
<View>
  <View
    style={
      {
        "marginVertical": 15,
      }
    }
    testID="loader"
  >
    <ActivityIndicator
      color="blue"
      size="small"
    />
  </View>
</View>
`;
